File: backend/app/schemas/dfd.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DFDNodeSchema(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str = Field(min_length=2, max_length=120)
    tipo: str = Field(pattern="^(proceso|almacen|entidad_externa)$")
    nombre: str = Field(min_length=1, max_length=255)
    descripcion: str = Field(default="", max_length=5000)
    numero_proceso: str | None = Field(default=None, max_length=40)
    ubicacion_proceso: str | None = Field(default=None, max_length=255)
    prefijo_almacen: str | None = Field(default=None, pattern="^(D|T|M)$")
    tipo_dato_almacen: str | None = Field(default=None, max_length=255)
    posicion_x: float
    posicion_y: float
    width: float | None = None
    height: float | None = None
    color: str | None = Field(default=None, max_length=32)
    fase: str | None = Field(default=None, max_length=120)
    categoria: str | None = Field(default=None, max_length=120)
    etiquetas: list[str] = Field(default_factory=list)


class DFDFlowSchema(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str = Field(min_length=2, max_length=120)
    origen_id: str = Field(min_length=2, max_length=120)
    destino_id: str = Field(min_length=2, max_length=120)
    etiqueta: str = Field(min_length=1, max_length=255)
    datos_descripcion: str | None = Field(default=None, max_length=5000)
    datos_campos: list[str] = Field(default_factory=list)
    fase: str | None = Field(default=None, max_length=120)
    tipo_flujo: str | None = Field(default=None, pattern="^(entrada|salida|bidireccional)$")
    tipo_relacion: str | None = Field(
        default=None,
        pattern="^(linea|flecha_abierta|flecha_cerrada|doble_flecha)$",
    )
    estilo_linea: str | None = Field(
        default=None,
        pattern="^(rectilinear|oblique|curve|round_oblique|round_rectilinear)$",
    )
    source_handle: str | None = Field(default=None, max_length=120)
    target_handle: str | None = Field(default=None, max_length=120)


class DFDCommentResponse(BaseModel):
    id: UUID
    model_id: UUID
    target_type: str
    target_client_id: str | None
    content: str
    status: str
    created_by_user_id: UUID | None
    created_by_user_email: str
    created_by_user_name: str | None
    created_by_user_type: str
    created_in_version_number: int
    created_at: datetime
    updated_at: datetime


class DFDCommentCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_type: str = Field(pattern="^(nodo|flujo)$")
    target_client_id: str | None = Field(default=None, max_length=120)
    content: str = Field(min_length=2, max_length=5000)


class DFDVersionResponse(BaseModel):
    id: UUID
    version_number: int
    created_at: datetime
    created_by_user_id: UUID | None
    created_by_user_email: str
    change_summary: str | None


class DFDVersionsResponse(BaseModel):
    model_id: UUID
    versions: list[DFDVersionResponse]


class DFDModelSnapshotRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    name: str = Field(min_length=1, max_length=255)
    description: str = Field(default="", max_length=10000)
    level: int = Field(ge=0, le=2)
    nodos: list[DFDNodeSchema] = Field(default_factory=list)
    flujos: list[DFDFlowSchema] = Field(default_factory=list)
    change_summary: str | None = Field(default=None, max_length=2000)


class DFDModelResponse(BaseModel):
    id: UUID
    project_id: UUID
    artifact_id: UUID
    phase: str
    name: str
    description: str
    level: int
    nodos: list[DFDNodeSchema]
    flujos: list[DFDFlowSchema]
    comentarios: list[DFDCommentResponse]
    version_actual: str
    current_version_number: int
    historial_versiones: list[DFDVersionResponse]
    created_at: datetime
    updated_at: datetime
    last_saved_at: datetime


class DFDVersionPreviewResponse(BaseModel):
    model_id: UUID
    source_version_number: int
    snapshot: DFDModelSnapshotRequest


class DFDRestoreVersionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    source_version_number: int = Field(ge=1)
    change_summary: str | None = Field(default=None, max_length=2000)

File: frontend/src/lib/api/dfd.ts
import { apiClient } from "@/lib/api/client";
import axios from "axios";
import type { ComentarioDFD, DiagramaFlujoDatos, FlujoDatos, NodoDFD, VersionDFD } from "@/lib/types/dfd.types";

type DFDCommentApi = {
  id: string;
  model_id: string;
  target_type: "nodo" | "flujo";
  target_client_id: string | null;
  content: string;
  status: "open" | "resolved";
  created_by_user_id: string | null;
  created_by_user_email: string;
  created_by_user_name: string | null;
  created_by_user_type: "CONSULTOR" | "EMPRESA" | "ADMINISTRADOR";
  created_in_version_number: number;
  created_at: string;
  updated_at: string;
};

type DFDVersionApi = {
  id: string;
  version_number: number;
  created_at: string;
  created_by_user_id: string | null;
  created_by_user_email: string;
  change_summary: string | null;
};

type DFDNodeApi = Omit<NodoDFD, "tipo"> & {
  tipo: "proceso" | "almacen" | "entidad_externa";
};

type DFDFlowApi = FlujoDatos;

type DFDModelApi = {
  id: string;
  project_id: string;
  artifact_id: string;
  phase: string;
  name: string;
  description: string;
  level: 0 | 1 | 2;
  nodos: DFDNodeApi[];
  flujos: DFDFlowApi[];
  comentarios: DFDCommentApi[];
  version_actual: string;
  current_version_number: number;
  historial_versiones: DFDVersionApi[];
  created_at: string;
  updated_at: string;
  last_saved_at: string;
};

type DFDVersionsApi = {
  model_id: string;
  versions: DFDVersionApi[];
};

type DFDPreviewApi = {
  model_id: string;
  source_version_number: number;
  snapshot: {
    name: string;
    description: string;
    level: 0 | 1 | 2;
    nodos: DFDNodeApi[];
    flujos: DFDFlowApi[];
    change_summary?: string | null;
  };
};

function mapCommentApiToUi(comment: DFDCommentApi): ComentarioDFD {
  return {
    id: comment.id,
    referencia_id: comment.target_client_id,
    referencia_tipo: comment.target_type,
    autor_id: comment.created_by_user_id ?? "unknown",
    autor_nombre: comment.created_by_user_name ?? comment.created_by_user_email.split("@")[0],
    autor_perfil: comment.created_by_user_type === "EMPRESA" ? "EMPRESA" : "CONSULTOR",
    contenido: comment.content,
    estado: comment.status === "resolved" ? "resuelto" : "abierto",
    created_at: comment.created_at,
  };
}

function mapVersionApiToUi(version: DFDVersionApi): VersionDFD {
  return {
    version: String(version.version_number),
    fecha: version.created_at,
    autor: version.created_by_user_email,
    descripcion_cambio: version.change_summary ?? "DFD update.",
  };
}

function mapModelApiToUi(model: DFDModelApi): DiagramaFlujoDatos {
  return {
    id: model.id,
    proyecto_id: model.project_id,
    entregable_id: model.artifact_id,
    nombre: model.name,
    descripcion: model.description,
    nivel: model.level,
    nodos: model.nodos,
    flujos: model.flujos,
    comentarios: model.comentarios.map(mapCommentApiToUi),
    version_actual: model.version_actual || String(model.current_version_number),
    historial_versiones: model.historial_versiones.map(mapVersionApiToUi),
    created_at: model.created_at,
    updated_at: model.updated_at,
  };
}

function mapNodeUiToApi(node: NodoDFD): DFDNodeApi {
  return {
    id: node.id,
    tipo: node.tipo,
    nombre: node.nombre,
    descripcion: node.descripcion,
    numero_proceso: node.numero_proceso,
    ubicacion_proceso: node.ubicacion_proceso,
    prefijo_almacen: node.prefijo_almacen,
    tipo_dato_almacen: node.tipo_dato_almacen,
    posicion_x: node.posicion_x,
    posicion_y: node.posicion_y,
    width: node.width,
    height: node.height,
    color: node.color,
    fase: node.fase,
    categoria: node.categoria,
    etiquetas: node.etiquetas ?? [],
  };
}

function mapFlowUiToApi(flow: FlujoDatos, includeRelation = true): DFDFlowApi {
  const mapped: DFDFlowApi = {
    id: flow.id,
    origen_id: flow.origen_id,
    destino_id: flow.destino_id,
    etiqueta: flow.etiqueta,
    datos_descripcion: flow.datos_descripcion,
    datos_campos: flow.datos_campos ?? [],
    fase: flow.fase,
    tipo_flujo: flow.tipo_flujo,
    estilo_linea: flow.estilo_linea,
    ...(flow.source_handle != null && { source_handle: flow.source_handle }),
    ...(flow.target_handle != null && { target_handle: flow.target_handle }),
  };

  if (includeRelation) {
    mapped.tipo_relacion = flow.tipo_relacion;
  }

  return mapped;
}

export const dfdApi = {
  async getModel(projectId: string, artifactId: string): Promise<DiagramaFlujoDatos> {
    const { data } = await apiClient.get<DFDModelApi>(`/projects/${projectId}/artifacts/${artifactId}/dfd`);
    return mapModelApiToUi(data);
  },

  async saveModel(
    projectId: string,
    artifactId: string,
    model: DiagramaFlujoDatos,
    changeSummary?: string
  ): Promise<DiagramaFlujoDatos> {
    const buildPayload = (includeRelation: boolean) => ({
      name: model.nombre || "DFD AS-IS",
      description: model.descripcion || "",
      level: model.nivel,
      nodos: model.nodos.map(mapNodeUiToApi),
      flujos: model.flujos.map((flow) => mapFlowUiToApi(flow, includeRelation)),
      change_summary: changeSummary ?? null,
    });

    const endpoint = `/projects/${projectId}/artifacts/${artifactId}/dfd`;

    try {
      const { data } = await apiClient.put<DFDModelApi>(endpoint, buildPayload(true));
      return mapModelApiToUi(data);
    } catch (error) {
      // Backward compatibility: some backend deployments still reject `tipo_relacion`.
      if (axios.isAxiosError(error) && error.response?.status === 422) {
        const { data } = await apiClient.put<DFDModelApi>(
          endpoint,
          buildPayload(false)
        );
        return mapModelApiToUi(data);
      }
      throw error;
    }
  },

  async getVersions(projectId: string, artifactId: string): Promise<VersionDFD[]> {
    const { data } = await apiClient.get<DFDVersionsApi>(
      `/projects/${projectId}/artifacts/${artifactId}/dfd/versions`
    );
    return data.versions.map(mapVersionApiToUi);
  },

  async createComment(
    projectId: string,
    artifactId: string,
    payload: { referencia_id: string | null; referencia_tipo: "nodo" | "flujo"; contenido: string }
  ): Promise<ComentarioDFD> {
    const { data } = await apiClient.post<DFDCommentApi>(`/projects/${projectId}/artifacts/${artifactId}/dfd/comments`, {
      target_type: payload.referencia_tipo,
      target_client_id: payload.referencia_id,
      content: payload.contenido,
    });
    return mapCommentApiToUi(data);
  },

  async previewVersion(
    projectId: string,
    artifactId: string,
    versionNumber: number,
    current: DiagramaFlujoDatos
  ): Promise<DiagramaFlujoDatos> {
    const { data } = await apiClient.get<DFDPreviewApi>(
      `/projects/${projectId}/artifacts/${artifactId}/dfd/versions/${versionNumber}/preview`
    );
    return {
      ...current,
      nombre: data.snapshot.name,
      descripcion: data.snapshot.description,
      nivel: data.snapshot.level,
      nodos: data.snapshot.nodos,
      flujos: data.snapshot.flujos,
      version_actual: String(versionNumber),
    };
  },

  async restoreVersion(
    projectId: string,
    artifactId: string,
    sourceVersionNumber: number,
    changeSummary?: string
  ): Promise<DiagramaFlujoDatos> {
    const { data } = await apiClient.post<DFDModelApi>(
      `/projects/${projectId}/artifacts/${artifactId}/dfd/restore-version`,
      {
        source_version_number: sourceVersionNumber,
        change_summary: changeSummary ?? null,
      }
    );
    return mapModelApiToUi(data);
  },
};
